### DISCUSSION TITLE
Questions about pcm options

### DISCUSSION THREAD

**Discussion by oleotiger** [opcm/pcm#253] [Q&A] [answered]
How to use the option `-m    | --multiple-instances | /m  => allow multiple PCM instances running in parallel`?
When to use and under what situation should I run pcm with `-m`?

In my own view, pcm.x sampels with PMU counters which can not be shared. That's why pcm.x and pcm-memory can not run at the same time. Why there is `-m`?

Does this option mean that I can run pcm.x and pcm-memory and pcm.power and so on simultaneously with `-m`?

**Comment by opcm** [comment]
-m is only for a scenario to run multiple pcm.x instances in parallel. The first pcm instance programs the PMUs and all others can detect that PMUs are already programmed by the first pcm instance and then only read the PMU without programming it.

**Comment by oleotiger** [comment]
Here is two more questions:
1. If `-m` only allows to run `pcm.x` in parallel, why should I run 2 or more pcm.x when just 1 instance can collect all the information I need? Is this option really useful?

2. As we all know, `perf` allows multiplexing sampling that PMU counters can be reused by time slots to collect metrics ( the number of metrics can be much more than the number of PUM counters). Does pcm tools support multiplexing? If yes, we can run pcm.x and pcm-memory simultaneously.

**Comment by opcm** [comment, marked as answer]
1. The multiple instances mode was mostly useful for reading PCM metrics *through API* from multiple processes/programs. The -m option in pcm.x was designed to test this functionality.

2. PCM tools have a simple design not allowing to miltiplex events from different processes. Within a process it is still possible to manually program and measure one event group after another frequently (for example pcm-pcie tool does that internally). However the multiplexing approach in general causes overhead because the PMUs need to be reprogrammed very frequently for non-steady state workloads.
